ca_cdk_constructs/edge_services/waf_v2_builder.py

```diff
@@ -0,0 +1,183 @@
+    add_custom_rule: Adds a custom rule to the WAFv2 WebACL.
+    add_managed_rule: Adds a managed rule to the WAFv2 WebACL.
+    build: Builds the WAFv2 WebACL.
+
+    Example:
+    ```python
+
+    waf_builder = WafV2Builder(
+      Stack(stack),
+      stage="test",
+      name="TestWaf",
+      description="A dummy WAF for testing",
+      tags={"Foo": "Bar"},
+    )
+
+    waf_builder.add_ip_rule(
+        name="TestIp",
+        priority=1,
+        addresses={"IPV4": ["17.0.0.0.1/24"]},
+        allow=True,
+        count_only=False,
+        cloud_watch_metrics_enabled=False,
+    )
+
+    waf = waf_builder.build()
+    """
+
+    def __init__(
+        self,
+        scope: Construct,
+        stage: str,
+        name: str,
+        description: str,
+        tags: dict = dict(),
+        waf_scope: str = "CLOUDFRONT",
+        log_group: cf_logs.LogGroup = None,
+        default_action: waf.CfnWebACL.DefaultActionProperty = None,
+    ) -> None:
+        self.rules = list()
+        self.scope = scope
+        self.stage = stage
+        self.name = name
+        self.waf_scope = waf_scope
+        self.description = description
+        self.log_group = log_group
+
+        if not default_action:
+            self.default_action = waf.CfnWebACL.DefaultActionProperty(
+                allow=waf.CfnWebACL.AllowActionProperty()
+            )
+
+        if log_group:
+            self.visibility_config = waf.CfnWebACL.VisibilityConfigProperty(
+                cloud_watch_metrics_enabled=True,
+                metric_name=self.name,
+                sampled_requests_enabled=True,
+            )
+        else:
+            self.visibility_config = waf.CfnWebACL.VisibilityConfigProperty(
+                cloud_watch_metrics_enabled=False,
+                metric_name=self.name,
+                sampled_requests_enabled=False,
+            )
+
+        self.tags = tags | {"Component": "WAF", "Stage": stage}
+
+    def add_custom_rule(self, rule: waf.CfnWebACL.RuleProperty) -> None:
+        self.rules.append(rule)
+
+    def add_managed_rule(
+        self,
+        name: str,
+        priority: int,
+        managed_rule_name: str,
+        managed_rule_vendor: str,
+        count_only: bool = False,
+        rules_to_exclude: list[str] = [],
+        cloud_watch_metrics_enabled: bool = False,
+    ) -> None:
+        """
+        Adds a managed rule to the WAFv2 WebACL.
+
+        :param name: The name of the rule.
+        :param priority: The priority of the rule.
+        :param managed_rule_name: The managed rule name.
+        :param managed_rule_vendor: The managed rule vendor.
+        :param count_only: Whether to only count the requests. Defaulse to False
+        :param rules_to_exclude: A list of rules to exclude. Defaults to an empty list.
+        :param cloud_watch_metrics_enabled: Whether to enable CloudWatch metrics. Defaults to False.
+        """
+        self.rules.append(
+            managed_rule_group_property(
```

Comment:
what's the benefit of keeping the functions in a separate file?

Reply:
Do you mean in a separate class? So they can be used independently of the builder if a more custom implementation of the WAF is required. I understand there are products that may require that, @ruthwells would know more.

Reply:
The builder is flexible enough. It can be used as a rule aggregator that can optionally create an ACL. 

```python
builder = WafV2Builder(......)

builder.add_managed_rule(.....)

builder.add_custom_rule(.....)

builder.add_ip_rule(.....)

# create my own ACL using the builder rules
my_acl = CfnWebACL(......, rules = builder.rules)

```

^^^ means the the stuff from `waf_rule_templates.py` can be moved inside the builder class

Reply:
if you do that ^^^ then it will be slightly confusing that you'd have to provide all the constructor arguments - they won't actually be used. But that's a minor thing